ngraph/core/reference/src/runtime/reference/random_uniform.cpp

```diff
@@ -0,0 +1,285 @@
+    memcpy(&x_bfloat16, &x_uint16, sizeof(x_uint16));
+    return x_bfloat16 - static_cast<bfloat16>(1);
+}
+
+// Runs Philox algorithm.
+void run_philox(uint64_t key, uint64_t counter, uint64_t n, size_t n_rounds, std::vector<uint32_t>& res) {
+    for (size_t i = 0; i < n_rounds; i++) {
+        calculate_round(key, counter, n);
+        if (i < n_rounds - 1)
+            raise_key(key);
+    }
+    auto res1 = split_high_low(n);
+    auto res2 = split_high_low(counter);
+    res[0] = res1.first;
+    res[1] = res1.second;
+    res[2] = res2.first;
+    res[3] = res2.second;
+}
+
+// Implementation of RandomUniform that uses Philox algorithm as inner random unsigned integer generator.
+void random_uniform(const uint64_t* out_shape,
+                    const char* min_val,
+                    const char* max_val,
+                    char* out,
+                    const Shape& out_shape_shape,
+                    const ngraph::element::Type& elem_type,
+                    uint64_t seed,
+                    uint64_t seed2) {
+    if (seed == 0 && seed2 == 0) {
+        std::srand(std::time(nullptr));
+        seed = std::rand();
+    }
+    uint64_t key = seed;
+    uint64_t counter = seed2;
+    uint64_t n = 0;
+    size_t shape_count = shape_size(out_shape_shape);
+    size_t elem_count = 1;
+    const size_t philox_output_size = 4;
+    for (size_t i = 0; i < shape_count; i++) {
+        elem_count *= out_shape[i];
+    }
+    // Each run of Philox algorithm generates 4 uint32 values.
+    // If output_type is int32, f32, bf16, or f16 each value is converted to
+    // corresponding type so we have 4 result values. For f64 and i64 we use
+    // a pair of values for conversion, so we have 2 result values.
+    // Step indicates how many values we generate in one iteration.
+    const size_t step = elem_type.size() > 4 ? 2 : 4;
+
+    for (size_t k = 0; k < elem_count; k += step) {
+        // generate 4 random uint32 values using Philox algorithm
+        std::vector<uint32_t> res(philox_output_size);
+        run_philox(key, counter, n, rounds_number, res);
+
+        // convert values to corresponding output_type
+        switch (elem_type) {
+        case ngraph::element::Type_t::f32: {
+            std::vector<float> res_float(step);
+            std::transform(res.data(), res.data() + step, res_float.data(), uint32_to_float);
+            float mn[1];
+            float mx[1];
+            memcpy(mn, min_val, elem_type.size());
+            memcpy(mx, max_val, elem_type.size());
+            // convert uint32 values to float32 and normalize to range
+            // [min_val, max_val)
+            std::transform(res.data(), res.data() + step, res_float.data(), [&mn, &mx](uint32_t elem) {
+                return uint32_to_float(elem) * (mx[0] - mn[0]) + mn[0];
+            });
+
+            memcpy(out + k * elem_type.size(), res_float.data(), std::min(step, elem_count - k) * elem_type.size());
+            break;
+        }
```

Comment:
can we use template instead because I see duplications here?

Reply:
I thought we try to avoid template code in nGraph?

Reply:
If we avoid template code but at the same moment duplicate it for different types it is the same.
We are avoiding template methods only in order to decrease impact to binary size, but the code duplication also impacts the foot print.

Reply:
Okay, I will change to template then.

Reply:
Done.